test_conformance/extensions/cl_khr_command_buffer/command_buffer_pipelined_enqueue.cpp

```diff
@@ -0,0 +1,307 @@
+        : BasicCommandBufferTest(device, context, queue)
+    {}
+
+    cl_int SetUpKernel() override
+    {
+        const char* mul_kernel_str =
+            R"(
+            __kernel void mul_by_val(int in, __global int* data)
+            {
+                size_t id = get_global_id(0);
+                data[id] *= in;
+            }
+
+            __kernel void increment(__global int* data)
+            {
+                size_t id = get_global_id(0);
+                data[id]++;
+            })";
+
+        cl_int error = create_single_kernel_helper_create_program(
+            context, &program, 1, &mul_kernel_str);
+        test_error(error, "Failed to create program with source");
+
+        error = clBuildProgram(program, 1, &device, nullptr, nullptr, nullptr);
+        test_error(error, "Failed to build program");
+
+        mul_kernel = clCreateKernel(program, "mul_by_val", &error);
+        test_error(error, "Failed to create mul_by_val kernel");
+
+        inc_kernel = clCreateKernel(program, "increment", &error);
+        test_error(error, "Failed to create increment kernel");
+
+        return CL_SUCCESS;
+    }
+
+    cl_int SetUpKernelArgs() override
+    {
+        cl_int error = CL_SUCCESS;
+        out_mem = clCreateBuffer(context, CL_MEM_READ_WRITE,
+                                 num_elements * buffer_size_multiplier
+                                     * sizeof(cl_int),
+                                 nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        cl_int val_arg = pattern;
+        error = clSetKernelArg(mul_kernel, 0, sizeof(cl_int), &val_arg);
+        test_error(error, "clSetKernelArg failed");
+
+        error = clSetKernelArg(mul_kernel, 1, sizeof(out_mem), &out_mem);
+        test_error(error, "clSetKernelArg failed");
+
+        error = clSetKernelArg(inc_kernel, 0, sizeof(out_mem), &out_mem);
+        test_error(error, "clSetKernelArg failed");
+
+        return CL_SUCCESS;
+    }
+
+    cl_int RecordCommandBuffer(clCommandBufferWrapper& cmd_buf)
+    {
+        cl_int error = clCommandNDRangeKernelKHR(
+            cmd_buf, nullptr, nullptr, inc_kernel, 1, nullptr, &num_elements,
+            nullptr, 0, nullptr, nullptr, nullptr);
+        test_error(error, "clCommandNDRangeKernelKHR failed");
+
+        error = clFinalizeCommandBufferKHR(cmd_buf);
+        test_error(error, "clFinalizeCommandBufferKHR failed");
+        return CL_SUCCESS;
+    }
+
+    const cl_int pattern = 42;
+
+    clKernelWrapper inc_kernel = nullptr;
+    clKernelWrapper mul_kernel = nullptr;
+};
+
+struct InOrderPipelined : public CommandBufferPipelined
+{
+    InOrderPipelined(cl_device_id device, cl_context context,
+                     cl_command_queue queue)
+        : CommandBufferPipelined(device, context, queue)
+    {}
+
+    cl_int Run() override
+    {
+        cl_int error = RecordCommandBuffer(command_buffer);
+        test_error(error, "RecordCommandBuffer failed");
+
```

Comment:
Hello, I believe the buffer needs to be initialised to 0, otherwise it will use whatever data it had and the test will fail because it is expecting 43.

Reply:
Thanks for catching that, I've pushed a commit that zero initialized the buffer in some setup code shared by all the pipeline tests.